openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,378 @@
+
+
+@intercept_errors(UserAPIInternalError, ignore_errors=[UserAPIRequestError])
+@transaction.commit_on_success
+def update_user_preferences(requesting_user, update, username=None):
+    """Update the user preferences for the given username.
+
+    Note:
+        It is up to the caller of this method to enforce the contract that this method is only called
+        with the user who made the request.
+
+    Arguments:
+        requesting_user (User): The user requesting to modify account information. Only the user with username
+            'username' has permissions to modify account information.
+        update (dict): The updated account field values.
+            Some notes:
+                Values are expected to be strings. Non-string values will be converted to strings.
+                Null values for a preference will be treated as a request to delete the key in question.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        UserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        UserNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        PreferenceValidationError: the update was not attempted because validation errors were found
+        PreferenceUpdateError: the operation failed when performing the update.
+        UserAPIInternalError: the operation failed due to an unexpected error.
+    """
+    existing_user = _get_user(requesting_user, username)
+
+    # First validate each preference setting
+    errors = {}
+    serializers = {}
+    for preference_key in update.keys():
+        preference_value = update[preference_key]
+        if preference_value is not None:
+            try:
+                serializer = _validate_user_preference(existing_user, preference_key, preference_value)
+                serializers[preference_key] = serializer
+            except PreferenceValidationError as error:
+                preference_error = error.preference_errors[preference_key]
+                errors[preference_key] = {
+                    "developer_message": preference_error["developer_message"],
+                    "user_message": preference_error["user_message"],
+                }
+    if errors:
+        raise PreferenceValidationError(errors)
+    # Then perform the patch
+    for preference_key in update.keys():
+        preference_value = update[preference_key]
+        if preference_value is not None:
+            try:
+                serializer = serializers[preference_key]
+                serializer.save()
+            except Exception as error:
+                raise _create_preference_update_error(preference_key, preference_value, error)
+        else:
+            delete_user_preference(requesting_user, preference_key)
```

Comment:
I had that originally, but ``delete_user_preference`` already catches exceptions and raises PreferenceUpdateError itself. The asymmetry in this code bothered me, but I didn't see a better way to do it.